executor/src/witgen/multiplicity_column_generator.rs

```diff
@@ -78,64 +65,68 @@ impl<'a, T: FieldElement> MultiplicityColumnGenerator<'a, T> {
             challenge_values: self.fixed.challenges.clone(),
         };
 
-        log::trace!(
-            "  Done building trace values, took: {}s",
-            start.elapsed().as_secs_f64()
-        );
-
-        for lookup in phantom_lookups {
-            log::trace!("  Updating multiplicity for phantom lookup: {lookup}");
-            let start = std::time::Instant::now();
-
-            let (rhs_machine_size, rhs_tuples) = self.get_tuples(&terminal_values, lookup.right);
-
-            log::trace!(
-                "    Done collecting RHS tuples, took {}s",
-                start.elapsed().as_secs_f64()
-            );
-            let start = std::time::Instant::now();
-
-            let index = rhs_tuples
-                .iter()
-                .map(|(i, tuple)| {
-                    // There might be multiple identical rows, but it's fine, we can pick any.
-                    (tuple, *i)
-                })
-                .collect::<HashMap<_, _>>();
-
-            log::trace!(
-                "    Done building index, took {}s",
-                start.elapsed().as_secs_f64()
-            );
-            let start = std::time::Instant::now();
+        // Index all bus receives with arbitrary multiplicity.
+        let receive_infos = self
+            .fixed
+            .bus_receives
+            .iter()
+            .filter(|(_, bus_receive)| {
+                bus_receive.has_arbitrary_multiplicity() && bus_receive.multiplicity.is_some()
+            })
+            .map(|(bus_id, bus_receive)| {
+                let (size, rhs_tuples) =
+                    self.get_tuples(&terminal_values, &bus_receive.selected_payload);
 
-            let (_, lhs_tuples) = self.get_tuples(&terminal_values, lookup.left);
+                let index = rhs_tuples
+                    .into_iter()
+                    .map(|(i, tuple)| {
+                        // There might be multiple identical rows, but it's fine, we can pick any.
+                        (tuple, i)
+                    })
+                    .collect::<HashMap<_, _>>();
+
+                (
+                    *bus_id,
+                    ReceiveInfo {
+                        multiplicity_column: match bus_receive.multiplicity.as_ref().unwrap() {
+                            AlgebraicExpression::Reference(r) => r.poly_id,
```

Comment:
Use `try_to_simple_poly`? Can it have a next ref?

Reply:
Right, good point!